src/main/java/org/jabref/model/entry/BibEntry.java

```diff
@@ -825,33 +830,46 @@ public String getUserComments() {
         }
     }
 
+    public KeywordList getFieldAsKeywords(Field field, Character keywordSeparator) {
+        Optional<KeywordList> storedList = fieldsAsKeywords.get(field, keywordSeparator);
```

Comment:
Can't you use [Google Guava's Cache](https://www.baeldung.com/guava-cacheloader)? 😇  Maybe, with [Eclipse Collections Twin](https://www.eclipse.org/collections/javadoc/9.1.0/org/eclipse/collections/api/tuple/Twin.html)

```java
LoadingCache<Pair<Field, Character>, KeywordList> storedKeywordList = CacheBuilder.newBuilder()
  .build(new CacheLoader<Pair<Field, Character>, KeywordList>() {
    @Override
    public String load(final Pair<Field, Character> pair) throws Exception {
      return getField(pair.getOne().map(... pair.getTwo());
    }
});
```

```java
return storedKeywordList.get(Tuples.pair(field, keywordSeparator));
```

OK, invalidation is worse (as it is not possible to invalidate all pairs having the same first value (easily)).

```java
storedKeywordList.invalidateAll();
```

(I now that Tuples.pair(...) might be slower than your approach. Howver, I would like that the cache can be tweaked ^^).

Reply:
That looks complicated 😄 https://github.com/google/guava/wiki/CachesExplained#applicability has the nice explanation
> Note: If you do not need the features of a Cache, ConcurrentHashMap is more memory-efficient -- but it is extremely difficult or impossible to duplicate most Cache features with any old ConcurrentMap.

I guess we do not need these cache features (but I've also never worked with Guave Caches...so I cannot appreciate them).